errors.go

```diff
@@ -0,0 +1,388 @@
+func IsAccessDenied(e error) bool {
+	type ad interface {
+		IsAccessDeniedError() bool
+	}
+	_, ok := Unwrap(e).(ad)
+	return ok
+}
+
+// ConvertSystemError converts system error to appropriate teleport error
+// if it is possible, otherwise, returns original error
+func ConvertSystemError(err error) error {
+	innerError := Unwrap(err)
+
+	if os.IsNotExist(innerError) {
+		return wrap(&NotFoundError{Message: innerError.Error()}, 2)
+	}
+	if os.IsPermission(innerError) {
+		return wrap(&AccessDeniedError{Message: innerError.Error()}, 2)
+	}
+	switch realErr := innerError.(type) {
+	case *net.OpError:
+		return wrap(&ConnectionProblemError{
+			Message: fmt.Sprintf("failed to connect to server %v", realErr.Addr),
+			Err:     realErr}, 2)
+	case *os.PathError:
+		return wrap(&AccessDeniedError{
+			Message: fmt.Sprintf("failed to execute command %v error:  %v", realErr.Path, realErr.Err),
+		}, 2)
+	case x509.SystemRootsError, x509.UnknownAuthorityError:
+		return wrap(&TrustError{Err: innerError}, 2)
+	default:
+		return err
+	}
+}
+
+// ConnectionProblem returns ConnectionProblem
+func ConnectionProblem(err error, message string, args ...interface{}) error {
+	return wrap(&ConnectionProblemError{
+		Message: fmt.Sprintf(message, args...),
+		Err:     err,
+	}, 2)
+}
+
+// ConnectionProblemError indicates any network error that has occured
+type ConnectionProblemError struct {
+	Message string `json:"message"`
+	Err     error  `json:"-"`
+}
+
+// Error is debug - friendly error message
+func (c *ConnectionProblemError) Error() string {
+	return fmt.Sprintf("%v: %v", c.Message, c.Err)
+}
+
+// IsConnectionProblemError indicates that this error is of ConnectionProblem
+func (c *ConnectionProblemError) IsConnectionProblemError() bool {
+	return true
+}
+
+// OrigError returns original error (in this case this is the error itself)
+func (c *ConnectionProblemError) OrigError() error {
+	return c
+}
+
+// IsConnectionProblem detects if this error is of ConnectionProblemError
+func IsConnectionProblem(e error) bool {
+	type ad interface {
+		IsConnectionProblemError() bool
+	}
+	_, ok := Unwrap(e).(ad)
+	return ok
+}
+
+// LimitExceeded returns new limit exceeded error
+func LimitExceeded(message string, args ...interface{}) error {
+	return wrap(&LimitExceededError{
+		Message: fmt.Sprintf(message, args...),
+	}, 2)
+}
+
+// LimitExceededError indicates rate limit or connection limit problem
+type LimitExceededError struct {
+	Message string `json:"message"`
+}
+
+// Error is debug - friendly error message
+func (c *LimitExceededError) Error() string {
+	return c.Message
+}
+
+// IsLimitExceededError indicates that this error is of ConnectionProblem
+func (c *LimitExceededError) IsLimitExceededError() bool {
+	return true
+}
+
+// OrigError returns original error (in this case this is the error itself)
+func (c *LimitExceededError) OrigError() error {
+	return c
+}
+
+// IsLimitExceeded detects if this error is of LimitExceededError
```

Comment:
`detects if` -> `returns whether` (for consistency - of other way around, i.e. swap every use with `detects`)